spec.bs

```diff
@@ -6254,6 +6225,210 @@ an [=origin=] |topLevelOrigin|, and a [=string=] |slotSizeQueryParam|:
+    |firstEntry|'s [=trusted scoring signals request/top level origin=].
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/seller script fetcher=] is equal to
+    |firstEntry|'s [=trusted scoring signals request/seller script fetcher=].
+  1. [=set/Append=] |entry|'s [=trusted scoring signals request/render URL=] to |renderURLs|.
+  1. [=list/Extend=] |adComponentRenderURLs| with |entry|'s [=trusted scoring signals request/
+    ad component URLs=].
+1. Return the result of [=building trusted scoring signals url=] with |firstEntry|'s [=trusted
+  scoring signals request/base URL=], |renderURLs|, |adComponentRenderURLs|, |firstEntry|'s
+  [=trusted scoring signals request/seller experiment group id=],
+  |firstEntry|'s [=trusted scoring signals request/top level origin=].
+
+</div>
+
+<div algorithm>
+To <dfn>check if trusted scoring signals batch honors URL length limit</dfn> given a
+[=trusted scoring signals batcher=] |batcher| and a non-empty [=list=] of [=trusted
+scoring signals requests=] |entriesToBatch|:
+
+1. If the [=list/size=] of |entriesToBatch| is 1, return true.
+1. If |batcher|'s [=trusted scoring signals batcher/url length limit=] is 0, return true.
+1. Return whether the [=string/length=] of the result of [=building batched trusted scoring signals
+  url=] for |entriesToBatch| is &le; |batcher|'s [=trusted scoring signals batcher/url length
+  limit=].
+
+</div>
+
+<div algorithm>
+
+To <dfn>batch and fetch trusted scoring signals</dfn> given a [=trusted scoring signals batcher=]
+|batcher|:
+
+1. Until this object is no longer needed:
+  1. Wait until at least one of the following is true:
+    1. |batcher|'s [=trusted scoring signals batcher/request queue=] [=map/is not empty=].
+    1. |batcher|'s [=trusted scoring signals batcher/request map=] [=map/is not empty=] and some
+      heuristically chosen amount of time has passed.
+  1. Atomically transfer all entries in |batcher|'s [=trusted scoring signals batcher/
+    request queue=] to |incomingRequests|, leaving |batcher|'s [=trusted scoring signals batcher/
+    request queue=] empty.
+  1. For each |request| in |incomingRequests|:
+    1. Let |key| be (|request|'s [=trusted scoring signals request/seller script fetcher=],
+      |request|'s [=trusted scoring signals request/base url=], |request|'s [=trusted scoring
+      signals request/seller experiment group id=], |request|'s [=trusted scoring signals request/
+      top level origin=]).
+    1. If |batcher|'s [=trusted scoring signals batcher/request map=][|key|] does not [=map/exist=],
+      then [=map/set=] |batcher|'s [=trusted scoring signals batcher/request map=][|key|] to an
+      empty [=list=].
+    1. [=list/Append=] |request| to |batcher|'s [=trusted scoring signals batcher/request map=]
+      [|key|].
+  1. Some number of times, heuristically, select a |key| and a non-empty sublist of |batcher|'s

```

Comment:
"some number of times", maybe "[=iteration/While=] |batcher|'s [=trusted scoring signals batcher/request map=] [=map/is not empty=]"?

Reply:
my understanding of the sentence here is:
```
for i in range(0, a heuristically chosen number) {
  select a key and a sublist;
  handle the batch;
}
```

where after the loop, the request map does not need to be empty. Is that correct?

Reply:
Yup, since it might want to wait a bit to collect more stuff in request map to get a bigger batch.

Hmm, I guess we technically are not requiring eventual progress here. Not immediately sure how to do so.

Reply:
maybe a note about this does not need to be drained during the for loop and can wait for more requests to batch, but just need to make sure all requests are batched (requestMap is empty) in the end ?

Reply:
Rephrased the note there somewhat (not pushed yet, though, working on Dominic's suggestion, too).